### PR TITLE
UniswapV2Bridge

### PR DESCRIPTION
**Pull request by xianny**
## Description

Implement UniswapV2Bridge contract. Allows us to swap tokens through the UniswapV2 Router. 

## Testing instructions

Please review unit tests and comment if you think any cases should be added. 

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        if (state.fromTokenAddress == toTokenAddress) { // NOT TESTED
+            LibERC20Token.transfer(state.fromTokenAddress, to, amount);
+            return BRIDGE_SUCCESS;
+        }
+
+        // Get our balance of `fromTokenAddress` token.
+        state.fromTokenBalance = IERC20Token(state.fromTokenAddress).balanceOf(address(this));
+
+        // Grant the Uniswap router an allowance.
+        LibERC20Token.approve(
```
Comment (dekz) [inline contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:82]: There should be a `LibERC20Token .approveIfBelow` so we keep it as high as possible and only write the state (expensive) if we have to.

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            amount,
+            // Convert `fromTokenAddress` to `toTokenAddress`.
+            path,
+            // Recipient is `to`.
+            to,
+            // Expires after this block.
+            block.timestamp
+        );
+
+        state.boughtAmount = amounts[0];
```
Comment (dekz) [inline contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:109]:
```

| amounts | uint[] memory | The input token amount and all subsequent output token amounts.|

```
https://uniswap.org/docs/v2/smart-contracts/router/#swapexacttokensfortokens

does this mean the first element is amountSold?

Comment (dorothy-zbornak) [reply contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:109]: Yeah seems like we want the second element here.

Comment (dorothy-zbornak) [reply contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:109]: Could also just do `state.boughtAmount = router.swapExactTokensForTokens(...)[1]`.

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+
+// solhint-disable space-after-comma
+// solhint-disable not-rely-on-time
+contract UniswapV2Bridge is
+    IERC20Bridge,
+    IWallet,
+    DeploymentConstants
+{
+    using LibAddressArray for address[];
```
Comment (dorothy-zbornak) [inline contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:39]: Think you don't need this.

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        LibERC20Token.approve(
+            state.fromTokenAddress,
+            _getUniswapV2Router01Address(),
+            state.fromTokenBalance
+        );
+
+        // Convert directly from fromTokenAddress to toTokenAddress
+        address[] memory path = new address[](2);
+        path[0] = state.fromTokenAddress;
+        path[1] = toTokenAddress;
```
Comment (dorothy-zbornak) [inline contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:91]: I thought there would be multihop support out of the gate? Should we just encode an `address[] path` as the bridge data?

Comment (xianny) [reply contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol:91]: Ya let's do that. I'll change it.

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol (hunk 1 above)
Comment (xianny) [reply contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol]: okiedokie will switch

File: contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol (hunk 2 above)
Comment (xianny) [reply contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol]: Oops my bad. Looks like we want the second element (last element if more than 2 tokens in the path).

**Pull request merged**